Assets/Features/Telemetry Lap File/IdealTelemetryLapCreator.cs

```diff
@@ -0,0 +1,238 @@
+            }
+
+            if (!AllEqual(m => m.fileFormatVersion, (x, y) => x == y))
+            {
+                throw new ArgumentException($"All laps must have the same file format");
+            }
+
+            if (!AllEqual(m => m.frequency, (x, y) => x == y))
+            {
+                throw new ArgumentException($"All laps must have the same frequency");
+            }
+
+            if (!AllEqual(m => m.trackName, (x, y) => x == y))
+            {
+                throw new ArgumentException($"All laps must be associated to the same track");
+            }
+
+            if (!AllEqual(m => m.headers, HeaderComparer))
+            {
+                throw new ArgumentException($"All laps must have the same headers");
+            }
+
+            if (!AllEqual(m => m.headerUnits, HeaderComparer))
+            {
+                throw new ArgumentException($"All laps must have the same header units");
+            }
+        }
+
+        private bool HeaderComparer(string [] a, string [] b)
+        {
+            if (a.Length != b.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < a.Length; i++)
+            {
+                if (a[i] != b[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool SectorComparer(float[] a, float [] b)
+        {
+            return a.Length == b.Length && a.Length > 0;
+        }
+
+        private bool AllEqual<T>(Func<TelemetryLapMetadata, T> getter, Func<T,T,bool> comparer)
+        {
+            TelemetryLapMetadata firstItem = metadatas[0];
+            Func<TelemetryLapMetadata, bool> predicate = metadata => comparer(getter(metadata), getter(firstItem)); // comparison with the first item 
+
+            return
+                metadatas
+                .Skip(1)
+                .All(predicate);            
+        }
+
+        private void ProcessMetadata(TelemetryLapMetadata telemetryLapMetadata)
+        {
+            lapToFilename[timeTable.LapCount] = telemetryLapMetadata.csvFile;
+            timeTable.AddLap(telemetryLapMetadata.sectorsTime);
+        }
+
+
+        private void WriteSectorInFile(int sector, string filename, TelemetryLapFileWriter telemetryLapFileWriter)
+        {
+            if (!filename.Contains("Telemetry"))
+            {
+                filename = Path.Combine("Telemetry", filename);
+            }
+            
+            if (!File.Exists(@filename))
+            {
+                Debug.LogError($"{filename} not found");
+                return;
+            }
+
+            //https://weblogs.asp.net/lorenh/236134
+            IEnumerable<string> enumerable = File.ReadLines(@filename);
+            IEnumerator<string> enumerator = enumerable.GetEnumerator();
+            enumerator.MoveNext();//skip headers
+            enumerator.MoveNext();//skip units
```

Comment:
Is there any reason why skip units here?

Reply:
The file has this format
```
FRAME,TIME,DISTANCE,TOTALTIME,TOTALDISTANCE,SEGMENTNUM,SECTOR,MARKERS,MARKERTIME,MARKERFLAG,POSITIONX,POSITIONY,POSITIONZ,ROTATIONX,ROTATIONY,ROTATIONZ,STEERINGANGLE,THROTTLE,BRAKEPRESSURE,GEAR,SPEED
n/a,s,m,s,m,n/a,n/a,n/a,n/a,n/a,m,m,m,°,°,°,°,%,bar,,km/h
11630.00000,-0.00026,-0.00909,23.26000,274.53280,1.00000,0.00000,0.00000,0.00000,0.00000,513.55960,139.29570,2356.75600,-2.24164,119.37270,-0.40631,82.84000,0.77481,0.00000,1.00000,35.53372
```
The file iterator starts pointing at the headers (first line). Line 208 skips the first row, the headers. Line 209 skips the units. Doing this, we can iterate safetly through the data because we know that the following lines are all data lines

Reply:
Thank you for this Manuel, this makes sense now! Thanks again.